Documentation: expansion of Monkeypatch to include mocked classes and dictionaries

EvanKepner opened the pull request:
Following up on #5250 I've added another expansion to the `monkeypatch` documentation:

1. Included the reference output from `pytest --fixtures` for `monkeypatch` to list methods.
2. All Python code-blocks are in `code-block` RST sections across the file.
3. Added conceptual scenarios for the major functions to the intro.
4. Modified the "simple example" to include broader explanatory comments.
5. Added "Monkeypatching returned objects: building mock classes" section.
6. Added "Monkeypatching dictionaries" section.
7. Applied changes from the `blacken-docs` `pre-commit` hook.

### PR Checklist

- [x] Target the `master` branch for bug fixes, documentation updates and trivial changes.
- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>`.

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
+All modifications will be undone after the requesting
+test function or fixture has finished. The ``raising``
+parameter determines if a ``KeyError`` or ``AttributeError``
+will be raised if the set/deletion operation has no target.
```
```suggestion
will be raised if the target of the set/deletion operation does not exist.
```

EvanKepner replied:
Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
+:py:meth:`monkeypatch.setenv` and :py:meth:`monkeypatch.delenv` can be used for
+these patches.
+
+4. Use :py:meth:`monkeypatch.syspath_prepend` to add to the system ``$PATH`` and
```
```suggestion
4. Use :py:meth:`monkeypatch.syspath_prepend` to modify ``sys.path`` safely, and
```

EvanKepner replied:
Applied in https://github.com/pytest-dev/pytest/pull/5315/commits/60e4fd448e0ed05bd71436e506fafd4643b0a105

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
+    def getssh():
+        """Simple function to return expanded homedir ssh path from $USER"""
+        user = os.getenv("$USER")
+        return os.path.join(os.path.expanduser(f"~{user}"), ".ssh")
```
To simplify this example, you can use:

```python
def getssh():
    """Simple function to return expanded homedir ssh path from $USER"""
    return os.path.expanduser("~/.ssh")
```

EvanKepner replied:
Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7 - this is a really good simplification. I was debating using `Path.home()` but the original example used `os` so I wanted to stick with it.

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
+always returns ``True``, or return different values from the ``json()`` mocked method
+based on input strings.
+
+This mock can be shared across tests using the ``fixture`` syntax:
```
```suggestion
This mock can be shared across tests using a ``fixture``:
```

A fixture is not really a new syntax. 😁

EvanKepner replied:
Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
+
+
+Furthermore, if the mock was designed to be applied to all tests, the ``fixture`` could
+be moved to ``conftest.py`` with ``autouse=True``.
```
```suggestion
be moved to a ``conftest.py`` file and use the ``autouse=True`` option.
```

EvanKepner replied:
Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -90,7 +266,7 @@ to do this using the ``setenv`` and ``delenv`` method. Our example code to test:
 .. code-block:: python
 
-    # contents of our original code file e.g. code.py
+    # Contents of our original code file e.g. code.py.
```
I would leave all those "contents -> Contents" changes lone, just for consistency with the rest of the docs.

EvanKepner replied:
Great catch, also ensured `monkeypatch` is lowercase everywhere except when it is the first word in the headers.

nicoddemus replied:
I still see comments starting with an upper case letter... would you mind changing all comments to start with lowercase? This is the pattern we use through the docs and code. 👍

blueyed replied:
> I still see comments starting with an upper case letter... would you mind changing all comments to start with lowercase? This is the pattern we use through the docs and code. 

That's new to me also - I see that there are comments without punctuation and starting lowercased, but when using punctuation / full sentences they should be capitalized (like above).

nicoddemus replied:
full sentences yes, but in one-liners like this we don't use punctuation and start with a lowercase letter (at least that's what I see browsing the docs quickly now)

blueyed replied:
Well, you will also see no typing hints currently - but maybe soon.. ;)
(what I want to say is that style can be improved, and I prefer it like above also)

nicoddemus replied:
Well I won't get into that discussion right now then, the current style is fine.

@EvanKepner could you just squash and rebase your commits then? Thanks!

EvanKepner replied:
No problem at all, squash/rebase coming soon. Also @nicoddemus sorry I misread your original comment and thought you wanted all comment lines capitalized but without punctuation unless part of a larger sentence :sweat_smile: 

I'm a fan of generally strict type hints and full doc-strings for each function. It's a difficult challenge with documentation because you don't want to add additional syntax that could confuse a reader. These tutorial examples are simplified to show how the patching works so I steered towards the least verbose form. 

Might be worth an update in the guidance to https://github.com/pytest-dev/pytest/blob/master/CONTRIBUTING.rst#write-documentation if you come to a consensus on standardized documentation style beyond the pre-commit checks for `black` in the docs. If you're seeking suggestions I like one sentence per line in RST with a hard-wrap (or just a hard-wrap at something sensible) to make suggested changes and PR reviews easier.

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -110,7 +286,7 @@ both paths can be safely tested without impacting the running environment:
 .. code-block:: python
 
-    # contents of our test file e.g. test_code.py
+    # Contents of our test file e.g. test_code.py.
```
Also, if I'm being pedantic, we usually don't use `.` at the end of single line comments

EvanKepner replied:
I changed this throughout the file with regard to the single line comments.

asottile commented on doc/en/monkeypatch.rst:
```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
+
+    def getssh():
+        """Simple function to return expanded homedir ssh path from $USER"""
+        user = os.getenv("$USER")
```
this looks up the environment variable named '$USER` -- I think you mean `os.getenv('USER')` (which could return `None`)

The `os.path.expanduser` call for "the current user" would just be:

```python
    return os.path.join(os.path.expanduser("~"), ".ssh")
```

no need for `getenv` or string formatting

nicoddemus commented on doc/en/monkeypatch.rst:
```diff
@@ -154,6 +330,112 @@ This behavior can be moved into ``fixture`` structures and shared across tests:
+            _ = app.create_connection_string()
+
+
+The modularity of the ``fixture`` syntax gives you the flexibility to define
```
```suggestion
The modularity of fixtures gives you the flexibility to define
```

EvanKepner replied:
Applied in https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7 - also realized the second test does not require the `mock_test_database` fixture so I removed it from the signature - https://github.com/pytest-dev/pytest/commit/7fa0e0405d4c051d176ba40ae4205dc3393e27a7#diff-425daef754438812ec1a0e4d3aeaeff8L434

asottile commented on doc/en/monkeypatch.rst:
```diff
@@ -8,46 +8,222 @@ Sometimes tests need to invoke functionality which depends
-directory, you can use the :py:meth:`monkeypatch.setattr` method to
-patch this function before calling into a function which uses it::
+Consider a scenario where you are working with user directories. In the context of
+testing, you do not want your test to depend on the running user. ``Monkeypatch``
```
I think we should consistently match the case of the fixture name, so

```rst
    ``monkeypatch``
```

instead of

```rst
    ``Monkeypatch``
```

EvanKepner commented:
@nicoddemus @asottile appreciate the fast review! I made the updates, agreed with all suggestions :+1:

asottile commented:
test failures here are expected -- until #5063 is merged

EvanKepner commented:
Squashed and rebased to `master` at https://github.com/pytest-dev/pytest/commit/ece774f0eb0bb87aa37c8c34df13a18be1500bfb. This includes the prior edits, and I went through and lower-cased all comment lines unless they were part of a sentence.

EvanKepner commented:
Realized the change to `os.path.expanduser("~/.ssh")` actually caused the test example to fail, since `expanduser` was patched to return `"/abc"`. Added a commit to update to this to use `pathlib` instead which is more clear and confirmed that it runs correctly.

EvanKepner commented:
@nicoddemus checking to make sure there is no need for additional action from me. If more edits are required please let me know. Thanks!

nicoddemus commented:
Hi @EvanKepner,

Thanks for the ping, missed that this was ready. 👍

EvanKepner commented:
Excellent - thanks! No worries at all :+1:
